lib/src/http_client_request.dart

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+// import 'dart:collection';
+import 'dart:convert';
+import 'dart:developer';
+import 'dart:ffi';
+import 'dart:io';
+import 'dart:isolate';
+import 'dart:typed_data';
+
+import 'package:ffi/ffi.dart';
+
+import 'exceptions.dart';
+
+import 'generated_bindings.dart';
+
+// TODO: Enable these when Headers are implemented
+// part '../third_party/http_headers.dart';
+// part '../third_party/http_date.dart';
+part 'http_client_response.dart';
+part 'http_callback_handler.dart';
+
+// Type definitions for various callbacks
+
+/// Called when a redirect is received. Gets new location and response code as
+/// arguments.
+typedef RedirectReceivedCallback = void Function(
+    String newLocationUrl, int responseCode);
+
+/// Called when server sends a response for the first time. Gets the response code
+/// as an argument.
+typedef ResponseStartedCallback = void Function(int responseCode);
+
+/// Called when a chunk of data is received from the server. Gets raw bytes as
+/// [List<int>], number of bytes read and response code as arguments.
+typedef ReadDataCallback = void Function(List<int> data, int bytesRead,
+    int responseCode); // onReadComplete may confuse people.
+
+/// Called when request is failed due to some reason. Gets [HttpException] as
+/// an argument which also contains the reason of failure.
+typedef FailedCallabck = void Function(HttpException exception);
+
+/// Called when a request is cancelled.
+typedef CanceledCallabck = void Function();
+
+/// Called when a request is finished with success. Gets the response code as
+/// an argument.
+typedef SuccessCallabck = void Function(int responseCode);
+
+/// HTTP request for a client connection.
+///
+/// It handles all of the Http Requests made by [HttpClient].
+/// Provides two ways to get data from the request.
+/// [registerCallbacks] or a [HttpClientResponse] which is a [Stream<List<int>>].
+/// Either of them can be used at a time.
+///
+/// Example Usage:
+/// ```dart
+/// final client = HttpClient();
+/// client.getUrl(Uri.parse('https://example.com/'))
+///   .then((HttpClientRequest request) {
+///   return request.close();
+/// }).then((HttpClientResponse response) {
+///   // Here you got the raw data.
+///   // Use it as you like.
+/// });
+/// ```
+// TODO: Implement other functions
+class HttpClientRequest implements IOSink {
+  final Uri _uri;
+  final String _method;
+  final Cronet _cronet;
+  final Pointer<Cronet_Engine> _cronetEngine;
+  final _CallbackHandler _cbh;
+  final Pointer<Cronet_UrlRequest> _request;
+  final Function _clientCleanup; // Holds the function to clean up after
```

Comment:
Move the comment before the field. And change it to a doc-comment so that you can easily see it in the IDE.